### PR TITLE
Support swaps with stablepool

### PR DESCRIPTION
**Pull request by JanKuczma**
The new `Router` contract will support swaps along paths that include `Pair` and `StablePool` types of pools.

### Swap Path
The path is a vector of `Step`s and must consist of at least 2 `Step`s. Each step is a struct `Step(PoolId, TokenId)`. The `TokenId` describes the token in of the swap for the given `PoolId`, except for the last `Step` - the last `Step` in the path always describes the token out. The `PoolId` of the last step is ignored.

E.g path `[(pool_1, token_a) , (pool_2, token_b) , (pool_x, token_c)]` can be visualized like this:
```shell
        pool_1          pool_2          pool_x - ignored
       /      \        /      \ 
token_a   ->   token_b    ->    token_c
```
### Compatibility
The old Router had a "pair caching" mechanism which allowed caching custom `Pair` contracts. Also, when adding liquidity to a non-existent pair, the pair was created through the `Factory` contract.

To keep these features, the new Router has to cache the StablePool contracts before they can be used for swaps.
When swapping, the Router checks for the pool in the following order:
1. Cached pools by `PoolId`
2. Cached pairs by `(token_in,token_out)`
3. Pairs created through Factory by `(token_in,token_out)`

### CODE REVIEW

**Review by deuszx — CHANGES_REQUESTED** [review]
Let's start with a general comment – I'd add a new router _alongside_ the old one, not replace its code. Call it however you want (maybe even `RouterV2`) but let's not replace the current code (including `traits`).

File: amm/contracts/router/lib.rs [hunk 1]
```diff
@@ -17,25 +18,27 @@ pub mod router {
...
-        pub fn new(factory: AccountId, wnative: AccountId) -> Self {
+        pub fn new(pair_factory: AccountId, wnative: AccountId) -> Self {
             Self {
-                factory,
+                pair_factory,
                 wnative,
                 owner: Self::env().caller(),
-                pairs: Default::default(),
+                cached_pairs: Default::default(),
+                cached_pools: Mapping::default(),
```
Comment (deuszx) [inline amm/contracts/router/lib.rs:41]: nit: let's make it consistent.

File: amm/contracts/router/lib.rs [hunk 2]
```diff
@@ -67,21 +75,48 @@ pub mod router {
...
             token_0: AccountId,
             token_1: AccountId,
         ) -> Result<(), CallerIsNotOwner> {
             ensure!(self.env().caller() == self.owner, CallerIsNotOwner);
-            self.pairs.remove((token_0, token_1));
-            self.pairs.remove((token_1, token_0));
+            if let Some((pair, _)) = self.cached_pairs.take((token_0, token_1)) {
+                self.cached_pools.remove(pair);
+            }
+            self.cached_pairs.remove((token_1, token_0));
```
Comment (deuszx) [inline amm/contracts/router/lib.rs:95]: We need to remove `(token_0, token_1)` as well.

File: amm/contracts/router/lib.rs [hunk 3]
```diff
@@ -67,21 +75,48 @@ pub mod router {
...
+        pub fn remove_pool_from_cache(
+            &mut self,
+            pool_id: AccountId,
+        ) -> Result<(), CallerIsNotOwner> {
+            ensure!(self.env().caller() == self.owner, CallerIsNotOwner);
+            if let Some(pool) = self.cached_pools.take(pool_id) {
+                match pool {
+                    Pool::Pair(tokens, _, _) => {
+                        self.cached_pairs.remove(tokens);
+                        self.cached_pairs.remove(tokens);
```
Comment (deuszx) [inline amm/contracts/router/lib.rs:109]: I think you need to destructure the tuple and remove twice – switching the order.

**Review by deuszx — CHANGES_REQUESTED** [review]
Some small errors around caching.

**Review by krzysztofziobro — COMMENTED** [review]
Haven't read the whole thing yet, first few things that caught my eye

File: amm/contracts/router/lib.rs (hunk 2 above)
Comment (JanKuczma) [reply amm/contracts/router/lib.rs:95]: I think `self.cached_pairs.take((token_0, token_1))` in line 92 already removes `(token_0, token_1)` :thinking:

Comment (deuszx) [reply amm/contracts/router/lib.rs:95]: It does.

File: amm/traits/router.rs [hunk 4]
```diff
@@ -185,45 +190,13 @@ pub trait Router {
     #[ink(message)]
     fn quote(&self, amount_0: u128, reserve_0: u128, reserve_1: u128) -> Result<u128, RouterError>;
```
[the remark this answers is not in the document]
Comment (JanKuczma) [reply amm/traits/router.rs:191]: Agree. I'm not sure if this method is used somewhere by FE but I doubt it (it does a simple calculation which can be done off-chain).

**Comment by JanKuczma** [comment]
The development continued in #117.

**Pull request closed without merging**